pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -516,11 +607,42 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(pInfo *framework.QueuedPodI
 	for plugin := range pInfo.UnschedulablePlugins {
 		metrics.UnschedulableReason(plugin, pInfo.Pod.Spec.SchedulerName).Inc()
 	}
-	if p.moveRequestCycle >= podSchedulingCycle {
+	// While the pod was in flight, a more recent cycle number might have
+	// been recorded for it in p.inFlightPods, so in that case we need to
+	// honor the more recent cycle number.
+	moveRequestCycle := pInfo.MoveRequestCycle
+	inFlightMoveRequestCycle := p.inFlightPods[pInfo.Pod.UID].moveRequestCycle
+	if inFlightMoveRequestCycle > moveRequestCycle {
+		moveRequestCycle = inFlightMoveRequestCycle
+	}
+
+	isWorthRequeue := false
+	for _, e := range p.inFlightPods[pInfo.Pod.UID].events {
```

Comment:
I am wondering... with your recent changes, we know exactly which events occurred while the pod was in flight.

Suppose all unschedulable plugins agree that the events they are waiting for are in this list. What happens now is that `isWorthRequeue` becomes true and the pod goes into the backoff queue.

Can't we go one step further and put the pod into the active queue instead in this case? That would be better because then scheduling doesn't get delayed by the backoff delay.

Conceptually it's the same as putting the pod into the unschedulable queue and then receiving the event that wakes it up.

Reply:
> Can't we go one step further and put the pod into the active queue instead in this case?

Good point. I think we can do the same as we do in movePodsToActiveOrBackoffQueue with PodImmediatelySchedulable.

One concern is that we likely need to look up more events and thus run more `Requeue()`s because we cannot `break` the loop unless we get PodImmediatelySchedulable. Given `Requeue` should be a lightweight one, I may be over-minding about it though.

Reply:
Done in https://github.com/kubernetes/kubernetes/pull/117844/commits/d467857ae0bdf60808bd11b2d200f7e55d73af59.

Could you check if it's what you meant? It contains many change for refactoring, but only behavior change is "to requeue in-flight Pod to activeQ if Requeue() return `PodImmediatelySchedulable` for any of events that happened during scheduling."

Reply:
I think we can do that also for `PodMaybeSchedulable`: if a pod is in the unschedulable queue, waiting for an event, and returns that for an incoming event, the pod would be moved to the active queue, right?

So skipping the backoff queue and going directly to the active queue also when such an event was missed while in flight makes sense to me.

Reply:
> I think we can do that also for PodMaybeSchedulable: if a pod is in the unschedulable queue, waiting for an event, and returns that for an incoming event, the pod would be moved to the active queue, right?

Currently No. For unschedulable Pods:
- PodMaybeSchedulable -> move a Pod to activeQ or backoffQ (honor backoff time.)
- PodImmediatelySchedulable -> move a Pod to activeQ (ignore backoff time.)

So, I think, for in-flight Pods as well, we need to honor backoff time if Requeue()s return `PodMaybeSchedulable` for events during scheduling.

Reply:
So we still need to handle `PodMaybeSchedulable` here because it might move the pod to the active queue? But I haven't looked at the code yet, perhaps you already do.

Reply:
Yes, correct. 
`requeuePodViaSchedulingHint` below moves this Pod to activeQ or backoffQ if some Requeue return `PodMaybeSchedulable` and no Requeue return `PodImmediatelySchedulable`.